File: whtml.js
Element.addMethods({
  appendTo: function(element, parent) {
    parent.appendChild(element);
  },
  
  setChildFunction: function(element, childFunction) {
    childFunction();
  },
  
  yield: function(element, context) {
    var positionNode = document.createTextNode('');
    element.appendChild(positionNode);
    context.apply(positionNode);
  }
});

var WHTML = {
  parts: {},
  customTags: {},
  
  initPage: function() {
    this.loadPart("page", document.getElementsByTagName('body')[0]);
  },
  
  loadPart: function(url, parent) {
    if(!WHTML.parts[url]) {
      new WHTML.Part(url);
    }
    WHTML.parts[url].whenLoaded(function() {
      this.create(parent);
    });
  },
  
  createElement: function(name) {
    if(WHTML.customTags[name]) {
      return new WHTML.customTags[name]();
    } else {
      return document.createElement(name);
    }
  },
  
  dynamicAttributeFor: function(target, name) {
    var attr = target.retrieve("dynamicAttribute_" + name);
    if(!attr) {
      attr = new WHTML.DynamicAttribute(target, name);
      target.store("dynamicAttribute_" + name, attr);
    }
    return attr;
  }
};

WHTML.Part = Class.create({
  initialize: function(url) {
    this.url = url;
    this.loaded = false;
    this.loadCallbacks = [];
    
    WHTML.parts[url] = this;

    var e = document.createElement("script");
    e.type = "text/javascript";
    e.src = this.url + ".js";
    document.getElementsByTagName('head')[0].appendChild(e);
  },

  whenLoaded: function(callback) {
    if(this.loaded) {
      callback.call(this);
    } else {
      this.loadCallbacks.push(callback);
    }
  },
  
  scriptLoaded: function() {
    this.loaded = true;
    this.loadCallbacks.invoke('call', this);
  }
});

WHTML.SimpleValue = Class.create({
  initialize: function(value) {
    this.value = value;
  },
  
  getValue: function() {
    return this.value;
  },

  addListener: function(listener) {},
  removeListener: function(listener) {},
});

WHTML.DynamicValue = Class.create({
  initialize: function(valueFunc, dependencies) {
    this.valueFunc = valueFunc;
    this.dependencies = dependencies;
  },
  
  getValue: function() {
    return this.valueFunc();
  },
  
  addListener: function(listener) {
    this.dependencies.invoke('addListener', listener);
  },
  
  removeListener: function(listener) {
    this.dependencies.invoke('removeListener', listener);
  }
});

WHTML.Dynamic = Class.create({
  initialize: function() {
    this.currentDependencies = [];
    this.depFunc = function() { return [this.dynValue]; };
  },
  
  updateDependencies: function() {
    this.currentDependencies.invoke('removeListener',this);
    this.currentDependencies = this.depFunc.call(this);
    this.currentDependencies.invoke('addListener', this);
  }
});

WHTML.DynamicAttribute = Class.create(WHTML.Dynamic, {
  initialize: function($super, target, name) {
    $super();
    this.target = target;
    this.name = name;
    this.valueFunc = null;
  },
  
  set: function(dynValue) {
    this.dynValue = dynValue;
    this.updateDependencies();
    this.update();
  },
  
  update: function() {
    this.target.writeAttribute(this.name, this.dynValue.getValue());
  }
});

WHTML.DynamicElement = Class.create(WHTML.Dynamic, {
  initialize: function($super, name, depFunc, childFunc) {
    $super();
    this.name = name;
    this.depFunc = depFunc;
    this.childFunc = childFunc;
    this.updateDependencies();
  },
  
  setChildFunction: function(childFunction) {
    //this.childFunc = childFunction;
  },
  
  appendTo: function(parent) {
    this.parent = parent;
    this.currentElement = document.createElement(this.name);
    this.childFunc(this.currentElement);
    this.parent.appendChild(this.currentElement);
  },
  
  update: function() {
    var newElement = document.createElement(this.name);
    this.childFunc(newElement);
    this.parent.replaceChild(newElement, this.currentElement);
    this.currentElement = newElement;
  }
});

WHTML.Dependency = Class.create({
  initialize: function() {
    this.listeners = [];
  },
  
  addListener: function(listener) {
    this.listeners.push(listener);
  },
  
  removeListener: function(listener) {
    this.listeners = this.listeners.without(listener);
  },
  
  update: function() {
    this.listeners.invoke('update');
  }
});

WHTML.Container = Class.create({
  initialize: function() {
    this.actionFunc = function() {};
    this.actions = null;
    this.isContainer = true;
  },
  
  apply: function(positionNode) {
    this.getActions().invoke('apply', positionNode);
  },
  
  revert: function(positionNode) {
    this.getActions().invoke('revert', positionNode);
  },
  
  getActions: function() {
    if(this.actions == null) {
      this.actions = [];
      this.actionFunc(this);
    }
    return this.actions;
  },
  
  appendChild: function(child) {
    this.actions.push(new WHTML.AppendChildAction(child));
  },
  
  writeAttribute: function(name, value) {
    this.actions.push(new WHTML.WriteAttributeAction(name, value));
  },
  
  yield: function(context) {
    this.actions.push(new WHTML.YieldAction(context));
  }
});

WHTML.CustomTag = Class.create(WHTML.Container, {
  initialize: function($super) {
    $super();
    this.attrDependencies = {};
  },
  
  writeAttribute: function(name, value) {
    this['attr_' + name] = new WHTML.SimpleValue(value);
    this.attributeChanged(name);
  },
  
  setChildFunction: function(childFunction) {
    this.actionFunc = childFunction;
  },
  
  store: function(name, value) {
    this[name] = value;
  },
  
  retrieve: function(name) {
    return this[name];
  },
  
  attributeDependencyFor: function(name) {
    if(!this.attrDependencies[name]) {
      this.attrDependencies[name] = new WHTML.Dependency([]);
    }
    return this.attrDependencies[name];
  },
  
  attributeChanged: function(name) {
    if(this.attrDependencies[name]) {
      this.attrDependencies[name].update();
    }
  }
});

WHTML.Case = Class.create(WHTML.Dynamic, {
  initialize: function($super, parent, dynValue) {
    $super();
    this.dynValue = dynValue;
    this.whenElements = [];
    this.currentWhen = null;
    this.positionNode = document.createTextNode('');
    parent.appendChild(this.positionNode);
  },
  
  when: function(condFunc, childFunc) {
    this.whenElements.push(new WHTML.When(condFunc, childFunc));
    this.updateDependencies();
    this.update();
  },
  
  update: function() {
    var value = this.dynValue.getValue();
    this.whenElements.each(function(whenElement) {
      if(whenElement.condFunc(value)) {
        if(whenElement == this.currentWhen) return;
        if(this.currentWhen) this.currentWhen.revert(this.positionNode);
        this.currentWhen = whenElement;
        this.currentWhen.apply(this.positionNode);
        return;
      }
    }, this);
  },
  
  appendChild: function(child) {
  }
});

WHTML.When = Class.create(WHTML.Container, {
  initialize: function($super, condFunc, actionFunc) {
    $super();
    this.condFunc = condFunc;
    this.actionFunc = actionFunc;
  },
});

WHTML.AppendChildAction = Class.create({
  initialize: function(child) {
    this.child = child;
  },
  
  apply: function(positionNode) {
    positionNode.parentNode.insertBefore(this.child, positionNode);
  },
  
  revert: function(positionNode) {
    positionNode.parentNode.removeChild(this.child);
  }
});

WHTML.WriteAttributeAction = Class.create({
  initialize: function(name, value) {
    this.name = name;
    this.value = value;
  },
  
  apply: function(positionNode) {
    this.oldValue = positionNode.parentNode.readAttribute(this.name);
    positionNode.parentNode.writeAttribute(this.name, this.value);
  },
  
  revert: function(positionNode) {
    positionNode.parentNode.writeAttribute(this.name, this.oldValue);
  }
});

WHTML.YieldAction = Class.create({
  initialize: function(context) {
    this.context = context;
  },
  
  apply: function(positionNode) {
    this.context.apply(positionNode);
  },
  
  revert: function(positionNode) {
    this.context.revert(positionNode);
  }
});

File: whtml.rb
require "nokogiri"

module WHTML
  class Processor
    class IndentingOutput
      def initialize(io)
        @io = io
        @indentation = 0
      end
      
      def puts(line)
        @io.puts(("  " * @indentation) + line)
      end
      
      def write(lines)
        lines.split("\n").each do |line|
          self.puts line.strip 
        end
      end
      
      def indent
        @indentation += 1
        yield
        @indentation -= 1
      end
      
      def write_block(header, footer)
        self.puts header if header
        self.indent do
          yield
        end
        self.puts footer if footer
      end
      
      def string
        @io.string
      end
    end
    
    class ProcessedValue
      attr_reader :js_code, :dependencies
      
      def initialize(value)
        @dependencies = []
        s = StringScanner.new value
        parts = []
        loop do
          s.scan /(.*?)@([\w]+)/m or break
          parts << process_text_part(s[1]) unless s[1].empty?
          parts << "context['attr_#{s[2]}'].getValue()"
          @dependencies << s[2]
        end
        parts << process_text_part(s.rest) unless s.rest.empty?
        @js_code = join_parts parts
      end
      
      def dependency_array
        "[#{@dependencies.map{ |d| "context.attributeDependencyFor('#{d}')" }.join ', '}]"
      end
    end
    
    class ProcessedStringValue < ProcessedValue
      def join_parts(parts)
        parts.join " + "
      end
      
      def process_text_part(part)
        "'#{Processor.escape_js_string part}'"
      end
      
      def to_dynamic_value
        if dependency_array.empty?
          "new WHTML.SimpleValue(#{js_code})"
        else
          "new WHTML.DynamicValue(function() { return #{js_code}; }, #{dependency_array})"
        end
      end
    end
    
    class ProcessedCodeValue < ProcessedValue
      def join_parts(parts)
        parts.join
      end
      
      def process_text_part(part)
        part
      end
    end
    
    def initialize
      @element_counter = 0
      @out = IndentingOutput.new StringIO.new
      @custom_tags = {}
    end
    
    def new_element_id(prefix)
      @element_counter += 1
      "#{prefix}#{@element_counter}"
    end
    
    def parse(source, part_name)
      doc = Nokogiri::XML.parse("<?xml version=\"1.0\"?><root xmlns:whtml=\"http://whtml.net/whtml\" xmlns:widgets=\"http://whtml.net/widgets\">#{source}</root>")
      
      parent_id = new_element_id "parent"
      @out.write_block "WHTML.parts['#{part_name}'].create = function(#{parent_id}) {", "}" do
        @out.puts "var context = this;"
        doc.root.children.each do |child|
          process_content_node child, parent_id
        end
      end
      
      @custom_tags.each do |name, node|
        attribute_names = (node["attributes"] || "").split(",")
        
        @out.write_block "WHTML.customTags['#{node["name"]}'] = Class.create(WHTML.CustomTag, {", "});" do
          attribute_names.each do |attr_name|
            @out.puts "get#{attr_name.capitalize}: function() { return this['attr_#{attr_name}'].getValue(); },"
            @out.puts "set#{attr_name.capitalize}: function(value) { this['attr_#{attr_name}'] = new WHTML.SimpleValue(value); this.attributeChanged('#{attr_name}'); },"
          end

          parent_id = new_element_id "parent"
          @out.write_block "appendTo: function(#{parent_id}) {", "}" do
            @out.puts "var context = this;"
            if node["type"] == "text/javascript"
              @out.puts "var parent = #{parent_id};"
              value = ProcessedCodeValue.new node.text
              @out.write value.js_code
            else
              node.children.each do |child|
                process_content_node child, parent_id
              end
            end
          end
        end
      end
      
      @out.puts "WHTML.parts['#{part_name}'].scriptLoaded();"
      @out.string
    end
    
    def process_content_node(node, parent)
      case node
      when Nokogiri::XML::Element
        if node.namespace and node.namespace.href == "http://whtml.net/whtml" and not @custom_tags["#{node.namespace.prefix}:#{node.name}"]
          case node.name
          when "customtag"
            @custom_tags[node["name"]] = node
          when "yield"
            @out.puts "#{parent}.yield(context);"
          when "region"
            parent_id = new_element_id "parent"
            @out.puts "var #{parent_id} = #{node['parent']};"
            node.children.each do |child|
              process_content_node child, parent_id
            end
          else
            puts "invalid tag: #{node.name}"
          end
        elsif node.name == "script" and (node["type"].nil? or node["type"] == "text/javascript")
          @out.puts "var parent = #{parent};"
          value = ProcessedCodeValue.new node.text
          @out.write value.js_code
        else
          if node.attributes["id"]
            element_id = "window.#{node.attributes["id"]}"
          else
            element_id = new_element_id node.name
            @out.puts "var #{element_id};"
          end

          attributes = node.attribute_nodes
          oncreate_attr = attributes.find { |attr| attr.name == "oncreate" and attr.namespace.href == "http://whtml.net/whtml" }
          oncreate_value = oncreate_attr && ProcessedCodeValue.new(oncreate_attr.value)
          
          element_creation = lambda do
            if oncreate_attr
              attributes.delete oncreate_attr
              @out.write "(function() { #{oncreate_value.js_code} }).call(#{element_id});"
            end
            
            attributes.each do |attr|
              if attr.name =~ /^on/
                value = ProcessedCodeValue.new attr.value
                @out.puts "Event.observe(#{element_id}, '#{attr.name[2..-1]}', function() { #{value.js_code} })"
              else
                value = ProcessedStringValue.new attr.value
                if value.dependencies.empty?
                  @out.puts "#{element_id}.writeAttribute('#{attr.name}', #{value.js_code});"
                else
                  @out.puts "WHTML.dynamicAttributeFor(#{element_id}, '#{attr.name}').set(#{value.to_dynamic_value});"
                end
              end
            end
          end
          
          if oncreate_attr.nil? or oncreate_value.dependencies.empty?
            @out.puts "#{element_id} = WHTML.createElement('#{node.namespace ? "#{node.namespace.prefix}:#{node.name}" : node.name}');"
            element_creation.call
          else
            @out.write_block "#{element_id} = new WHTML.DynamicElement('#{node.name}', function() { return #{oncreate_value.dependency_array}; }, function(#{element_id}) {", "});" do
              element_creation.call
            end
          end

          unless node.children.empty?
            @out.write_block "#{element_id}.setChildFunction(function() {", "});" do
              node.children.each do |child|
                process_content_node child, element_id
              end
            end
          end
          
          @out.puts "#{element_id}.appendTo(#{parent});"
        end
      when Nokogiri::XML::Text
        @out.puts "#{parent}.appendChild(document.createTextNode('#{Processor.escape_js_string node.text}'));"
      when Nokogiri::XML::Comment
        # ignore
      else
        raise RuntimeError
      end
    end
    
    def self.escape_js_string(string)
      string = string.dup
      string.gsub! "\n", "\\n"
      string.gsub! "'", "\\\\'"
      string
    end
  end
end